### PR TITLE
Simplification of core

### PR DESCRIPTION
**Pull request by piersy**
This is a draft PR, for the benefit of @aneequesafdar, its a long way from being ready so I wouldn't advise reviewing it at present.

### CODE REVIEW

File: common/keygenerator/generator_test.go [hunk 1]
```diff
@@ -10,3 +14,21 @@ func TestGenerator_Next(t *testing.T) {
...
+}
+
+func BenchmarkEcdsaGenerate(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		_, err := crypto.GenerateKey()
+		if err != nil {
+			b.Fatal(err)
+		}
+	}
+}
```
Comment (aneequesafdar) [inline common/keygenerator/generator_test.go:34]: I feel as though this is experimental code so either it should be removed or moved to a different branch

File: consensus/tendermint/algorithm/tendermint.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+import (
+	"encoding/hex"
+	"fmt"
+)
+
+type ValueID [32]byte
+
+func (v ValueID) String() string {
+	return hex.EncodeToString(v[:3])
+}
```
Comment (aneequesafdar) [inline consensus/tendermint/algorithm/tendermint.go:12]: Why do you want to only print the first three bytes of ValueID?

File: consensus/tendermint/algorithm/tendermint.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+type ConsensusMessage struct {
+	MsgType    Step
+	Height     uint64
+	Round      int64
+	Value      ValueID
+	ValidRound int64
+}
+
+func (cm *ConsensusMessage) String() string {
+	return fmt.Sprintf("s:%-3s h:%-3d r:%-3d v:%-6s", cm.MsgType.ShortString(), cm.Height, cm.Round, cm.Value.String())
```
Comment (aneequesafdar) [inline consensus/tendermint/algorithm/tendermint.go:81]: Just to confirm `%-3` and `%-6" is padding 3 and 6 spaces respectively to the right?

File: consensus/tendermint/algorithm/tendermint.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+// Oracle is used to answer questions the algorithm may have about its
+// state, such as 'Am I the proposer' or 'Have i reached prevote quorum
+// threshold for value with id v?'
+type Oracle interface {
+	Valid(ValueID) bool
+	MatchingProposal(*ConsensusMessage) *ConsensusMessage
+	PrevoteQThresh(round int64, value *ValueID) bool
+	PrecommitQThresh(round int64, value *ValueID) bool
```
Comment (aneequesafdar) [inline consensus/tendermint/algorithm/tendermint.go:91]: Do we need to have separate functions for Prevote and Precommit Quroum threshold, when we know both thresholds will always be the same?

File: consensus/tendermint/algorithm/tendermint.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+	// FThresh indicates whether we have messages whose voting power exceeds
+	// the failure threshold for the given round.
+	FThresh(round int64) bool
+	Proposer(round int64, nodeID NodeID) bool
```
Comment (aneequesafdar) [inline consensus/tendermint/algorithm/tendermint.go:95]: I feel as though height also needs to be provided to the implementation of oracle for it to correctly determine the proposer with nodeID is the proposer for the current height and round

File: common/keygenerator/generator_test.go (hunk 1 above)
Comment (piersy) [reply common/keygenerator/generator_test.go:34]: Yep

File: consensus/tendermint/algorithm/tendermint.go (hunk 2 above)
Comment (piersy) [reply consensus/tendermint/algorithm/tendermint.go:12]: For readability of logs. 3 bytes gives enough entropy to be very sure that we won't have collisions for the kind of network sizes that we will be testing and debugging.

**Comment by aneequesafdar** [comment]
closing in favour of #759

**Pull request closed without merging**